File: NeetCode/09-Backtracking/Permutations.cpp
#include <vector>
#include <iostream>
#include <algorithm>
using namespace std;
class Solution {
    public:
        vector<vector<int>> permute(vector<int>& nums) {
            vector<vector<int>> result;
            vector<int> currentList;
            vector<bool> used(nums.size(), false);
            backtrack(nums, currentList, result, used);
            return result;
        }
        void backtrack(vector<int>& nums, vector<int>& currentList, vector<vector<int>>& result, vector<bool>& used){
            if(currentList.size() == nums.size()){
                result.push_back(currentList);
                return;
            }
            
            for(int i = 0;i<nums.size();i++){
                if(used[i] == true){
                    continue;
                }
                used[i] = true;
                currentList.push_back(nums[i]);
                backtrack(nums, currentList, result, used);
                currentList.pop_back();
                used[i] = false;
            }
        }
    };

File: NeetCode/09-Backtracking/Permutations.java
import java.util.ArrayList;
import java.util.List;

public class Permutations {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        boolean[] used = new boolean[nums.length];
        backtrack(nums, new ArrayList<>(), result, used);
        return result;
    }
    private void backtrack(int[] nums, List<Integer> currentList, List<List<Integer>> result, boolean[] used){
        if(currentList.size() == nums.length){
            result.add(new ArrayList<>(currentList));
            return;
        }
        
        for(int i = 0;i<nums.length;i++){
            if(used[i] == true){
                continue;
            }
            used[i] = true;
            currentList.add(nums[i]);
            backtrack(nums, currentList, result, used);
            currentList.remove(currentList.size()-1);
            used[i] = false;
        }
    }
}
